File: cpp/src/Ice/ObjectFactoryManager.cpp

// **********************************************************************
//
// Copyright (c) 2003-2016 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

#include <Ice/ObjectFactoryManager.h>
#include <Ice/ObjectFactory.h>
#include <Ice/Functional.h>
#include <Ice/LocalException.h>
using namespace std;
using namespace Ice;
using namespace IceInternal;

IceUtil::Shared* IceInternal::upCast(ObjectFactoryManager* p) { return p; }

void
IceInternal::ObjectFactoryManager::add(const ObjectFactoryPtr& factory, const string& id)
{
    IceUtil::Mutex::Lock sync(*this);

    if((_factoryMapHint != _factoryMap.end() && _factoryMapHint->first == id)
       || _factoryMap.find(id) != _factoryMap.end())
    {
        AlreadyRegisteredException ex(__FILE__, __LINE__);
        ex.kindOfObject = "object factory";
        ex.id = id;
        throw ex;
    }

    _factoryMapHint = _factoryMap.insert(_factoryMapHint, pair<const string, ObjectFactoryPtr>(id, factory));
}

void
IceInternal::ObjectFactoryManager::remove(const string& id)
{
    ObjectFactoryPtr factory = 0;
    {
        IceUtil::Mutex::Lock sync(*this);

        FactoryMap::iterator p = _factoryMap.end();
        if(_factoryMapHint != _factoryMap.end())
        {
            if(_factoryMapHint->first == id)
            {
                p = _factoryMapHint;
            }
        }

        if(p == _factoryMap.end())
        {
            p = _factoryMap.find(id);
            if(p == _factoryMap.end())
            {
                NotRegisteredException ex(__FILE__, __LINE__);
                ex.kindOfObject = "object factory";
                ex.id = id;
                throw ex;
            }
        }
        assert(p != _factoryMap.end());

        factory = p->second;

        if(p == _factoryMapHint)
        {
            _factoryMap.erase(p++);
            _factoryMapHint = p;
        }
        else
        {
            _factoryMap.erase(p);
        }
    }

    //
    // Destroy outside the lock
    //
    assert(factory != 0);
    factory->destroy();
}

ObjectFactoryPtr
IceInternal::ObjectFactoryManager::find(const string& id) const
{
    IceUtil::Mutex::Lock sync(*this);

    FactoryMap& factoryMap = const_cast<FactoryMap&>(_factoryMap);

    FactoryMap::iterator p = factoryMap.end();
    if(_factoryMapHint != factoryMap.end())
    {
        if(_factoryMapHint->first == id)
        {
            p = _factoryMapHint;
        }
    }

    if(p == factoryMap.end())
    {
        p = factoryMap.find(id);
    }

    if(p != factoryMap.end())
    {
        _factoryMapHint = p;
        return p->second;
    }
    else
    {
        return 0;
    }
}

IceInternal::ObjectFactoryManager::ObjectFactoryManager() :
    _factoryMapHint(_factoryMap.end())
{
}

void
IceInternal::ObjectFactoryManager::destroy()
{
    FactoryMap oldMap;
    {
        IceUtil::Mutex::Lock sync(*this);
        oldMap.swap(_factoryMap);
        _factoryMapHint = _factoryMap.end();
    }

    //
    // Destroy all outside lock
    //
    for_each(oldMap.begin(), oldMap.end(),
             Ice::secondVoidMemFun<const string, ObjectFactory>(&ObjectFactory::destroy));
}

File: csharp/src/Ice/ObjectFactoryManager.cs
// **********************************************************************
//
// Copyright (c) 2003-2018 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

namespace IceInternal
{

    using System.Collections.Generic;

    public sealed class ObjectFactoryManager
    {
        public void add(Ice.ObjectFactory factory, string id)
        {
            lock(this)
            {
                if(_factoryMap.ContainsKey(id))
                {
                    Ice.AlreadyRegisteredException ex = new Ice.AlreadyRegisteredException();
                    ex.id = id;
                    ex.kindOfObject = "object factory";
                    throw ex;
                }
                _factoryMap[id] = factory;
            }
        }
        
        public void remove(string id)
        {
            object o = null;
            lock(this)
            {
                if(!_factoryMap.ContainsKey(id))
                {
                    Ice.NotRegisteredException ex = new Ice.NotRegisteredException();
                    ex.id = id;
                    ex.kindOfObject = "object factory";
                    throw ex;
                }
                _factoryMap.Remove(id);
            }
            ((Ice.ObjectFactory)o).destroy();
        }
        
        public Ice.ObjectFactory find(string id)
        {
            lock(this)
            {
                Ice.ObjectFactory factory = null;
                _factoryMap.TryGetValue(id, out factory);
                return factory;
            }
        }
        
        //
        // Only for use by Instance
        //
        internal ObjectFactoryManager()
        {
            _factoryMap = new Dictionary<string, Ice.ObjectFactory>();
        }
        
        internal void destroy()
        {
            Dictionary<string, Ice.ObjectFactory> oldMap = null;

            lock(this)
            {
                oldMap = _factoryMap;
                _factoryMap = new Dictionary<string, Ice.ObjectFactory>();
            }

            foreach(Ice.ObjectFactory factory in oldMap.Values)
            {
                factory.destroy();
            }
        }
        
        private Dictionary<string, Ice.ObjectFactory> _factoryMap;
    }

}
